feat: Added POST endpoint for Users

zxkevin opened the pull request:
Added a POST endpoint for users to check if
username, first name, or last name is valid
Check for imageUrl, if valid use
If no imageUrl default imageUrl will be used on creation in database

#62

zxkevin committed:
feat: Added POST endpoint for Users

Added a POST endpoint for users to check if
username, first name, or last name is valid
Check for imageUrl, if valid use
If no imageUrl default imageUrl will be used on creation in database

#62

3dxp commented on packages/hackmcx-api/src/controllers/users.controller.js:
```diff
@@ -21,3 +21,46 @@ export async function getUsersByUserId(req, res){
+    const noGivenImageUrl = !req.body.imageUrl;
+    const notPicture = !req.body.imageUrl || !/(https?:\/\/.*\.(?:png|jpg|gif|svg))/i.test(req.body.imageUrl);
+
+    if(emptyUsername){
```
We want to do all the validation checks we can before making a database query for this short circuit style evaluation. This can save a trip to the DB when the request is invalid anyway.

3dxp commented on packages/hackmcx-api/src/controllers/users.controller.js:
```diff
@@ -21,3 +21,46 @@ export async function getUsersByUserId(req, res){
+        res.send({error: "Invalid Username: Username is already taken."});
+        return
+    }
+    if(!noGivenImageUrl && notPicture){
```
I think we might want to take another look at this error message. If we are providing a default, then a "missing" user name does not invalidate a request.

3dxp commented on packages/hackmcx-api/src/routers/v1/users.router.js:
```diff
@@ -1,7 +1,8 @@
 
 router.get('/', getUsers)
 router.get('/:userId', getUsersByUserId)
+router.post('/', postUser)
```
This is just a style preference, but I try to group paths together. So anything with the '/' path in one code block, etc... Not anything that will prevent the PR from being merged, but just keeps things more consistent with the structure of other routers.

3dxp commented on packages/hackmcx-api/src/controllers/users.controller.js:
```diff
@@ -21,3 +21,46 @@ export async function getUsersByUserId(req, res){
+    const emptyUsername = !req.body.username || /^\s*$/.test(req.body.username);
+    const emptyFirstName = !req.body.firstname || /^\s*$/.test(req.body.firstname);
+    const emptyLastName = !req.body.lastname || /^\s*$/.test(req.body.lastname);
+    const usernameExists = (await dbClient.select('username').from('users').where('username', req.body.username)).length > 0 ? true : false;
```
We might be able to look for an exception on insert instead of doing an explicit check here just to reduce the total number of DB calls.

zxkevin committed:
fix: Removed select query for username validation

Changed the username validation so a try catch block
Updated imageUrl error message

#62

3dxp commented on packages/hackmcx-api/src/controllers/users.controller.js:
```diff
@@ -46,21 +44,21 @@ export async function postUser(req, res){
+        res.header('Location',`${req.baseUrl}/${id}` );
+        res.send();
+    }catch(e){
+        res.statusCode = 400;
```
I don't think we should assume any error is a PK conflict, we might want to check the error message to make sure in this case, and return a 400 if there is a PK conflict, and maybe something different otherwise. 

Eventually we might want to look at a global error handler to allow the app to return 5XX errors instead of crashing.

zxkevin committed:
fix: Update catch clause to validate correct error

Added a check to make sure the correct response is given for
the error in the try catch block for user sql insert

#62

3dxp reviewed (approved):
LGTM. Nice work on this!

acampb65 reviewed (approved):
Tested locally. Looks good to me!
